### PR TITLE
internal/radix51: implement (*FieldElement).Mul32

### PR DESCRIPTION
**Pull request by FiloSottile**
This pure Go implementation of Mul32 is more than twice as fast as the
assembly Mul implementation, and four times faster than the pure Go Mul.

```
Mul32          7.91ns ± 1%
Mul            18.6ns ± 1%
Mul [purego]   33.4ns ± 0%
```

Before Go 1.13, where we can't use math/bits because the fallbacks might
not be constant time, Mul32 is a little slower, but not nearly as much
as the pure Go Mul.

```
Mul32          9.74ns ± 0%
Mul [purego]   75.4ns ± 1%
```

/cc @hdevalence

### CODE REVIEW

File: internal/radix51/mul_bits.go [hunk 1]
```diff
@@ -1,18 +1,28 @@
 // Copyright (c) 2019 George Tankersley. All rights reserved.
+// Copyright (c) 2019 The Go Authors. All rights reserved.
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build go1.12
+// +build go1.13
```
Comment (gtank) [inline internal/radix51/mul_bits.go]: This kicks go1.12 over to `_compat` even though it has the `bits` package.

File: internal/radix51/mul_compat.go [hunk 2]
```diff
@@ -1,14 +1,16 @@
 // Copyright (c) 2017 George Tankersley. All rights reserved.
+// Copyright (c) 2019 The Go Authors. All rights reserved.
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build !go1.12
+// +build !go1.13
```
Comment (gtank) [inline internal/radix51/mul_compat.go]: Same as above

File: internal/radix51/bench_test.go [hunk 3]
```diff
@@ -29,3 +29,12 @@ func BenchmarkMul(b *testing.B) {
 		x.Mul(&x, &y)
 	}
 }
+
+func BenchmarkMul32(b *testing.B) {
+	var x radix51.FieldElement
+	x.One()
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		x.Mul32(&x, 0b10101010_10101010_10101010_10101010)
```
Comment (gtank) [inline internal/radix51/bench_test.go]: This is invalid syntax prior to go1.13 but the module supports anything from go1.12. Integer and a comment?

Aside: we actually support older versions than that via the `_compat` implementations. Should we change the module back to at least 1.11?

File: internal/radix51/mul_bits.go (hunk 1 above)
Comment (gtank) [reply internal/radix51/mul_bits.go]: Ah, I see. `bits` was vartime on platforms without the intrinsics?

Comment (FiloSottile) [reply internal/radix51/mul_bits.go]: > Ah, I see. `bits` was vartime on platforms without the intrinsics?

Precisely. And now if people want performance they can just run the latest Go.

File: internal/radix51/bench_test.go (hunk 3 above)
Comment (FiloSottile) [reply internal/radix51/bench_test.go]: I figured we didn't need as much compatibility in tests, but that argument doesn't make that much sense now that I think about it.

However, why is this working when the go.mod read go1.12...

**Comment by FiloSottile** [comment]
Updated the go.mod version and removed the clever binary syntax.

**Pull request merged**